src/annotator/util/buckets.js

```diff
@@ -88,161 +95,132 @@ export function findClosestOffscreenAnchor(anchors, direction) {
+ * by top offset
  *
  * @param {Anchor[]} anchors
- * @return {PositionPoints}
+ * @return {AnchorPosition[]}
  */
-export function constructPositionPoints(anchors) {
-  const aboveScreenAnchors = new Set();
-  const belowScreenAnchors = new Set();
-  const points = /** @type {PositionPoint[]} */ ([]);
+function getAnchorPositions(anchors) {
+  const anchorPositions = [];
 
-  for (let anchor of anchors) {
+  anchors.forEach(anchor => {
     if (!anchor.highlights?.length) {
-      continue;
+      return;
     }
+    const anchorBox = getBoundingClientRect(anchor.highlights);
+    anchorPositions.push({
+      top: anchorBox.top,
+      bottom: anchorBox.bottom,
+      anchor,
+    });
+  });
 
-    const rect = getBoundingClientRect(anchor.highlights);
-
-    if (rect.top < BUCKET_TOP_THRESHOLD) {
-      aboveScreenAnchors.add(anchor);
-    } else if (rect.top > window.innerHeight - BUCKET_NAV_SIZE) {
-      belowScreenAnchors.add(anchor);
-    } else {
-      // Add a point for the top of this anchor's highlight box
-      points.push([rect.top, 1, anchor]);
-      // Add a point for the bottom of this anchor's highlight box
-      points.push([rect.bottom, -1, anchor]);
-    }
-  }
-
-  // Sort onscreen points by pixel position, secondarily by position "type"
-  // (top or bottom of higlight box)
-  points.sort((a, b) => {
-    for (let i = 0; i < a.length; i++) {
-      if (a[i] < b[i]) {
-        return -1;
-      } else if (a[i] > b[i]) {
-        return 1;
-      }
+  // Now sort by top position
+  anchorPositions.sort((a, b) => {
+    if (a.top < b.top) {
+      return -1;
     }
-    return 0;
+    return 1;
   });
 
-  return {
-    above: Array.from(aboveScreenAnchors),
-    below: Array.from(belowScreenAnchors),
-    points,
-  };
+  return anchorPositions;
 }
 
 /**
- * Take a sorted set of `points` representing top and bottom positions of anchor
- * highlights and group them into a collection of "buckets".
+ * Compute buckets
  *
- * @param {PositionPoint[]} points
+ * @param {Anchor[]} anchors
  * @return {Bucket[]}
  */
-export function buildBuckets(points) {
+export function anchorBuckets(anchors) {
+  const anchorPositions = getAnchorPositions(anchors);
+  const aboveScreen = new Set();
+  const belowScreen = new Set();
   const buckets = /** @type {Bucket[]} */ ([]);
 
-  // Anchors that are part of the currently-being-built bucket, and a correspon-
-  // ding count of unclosed top edges seen for that anchor
-  const current = /** @type {{anchors: Anchor[], counts: number[] }} */ ({
-    anchors: [],
-    counts: [],
-  });
+  // Hold current working anchors and positions as we build each bucket
+  /** @type {WorkingBucket|null} */
+  let currentBucket = null;
+
+  /**
+   * Create a new working bucket based on the provided `AnchorPosition`
+   *
+   * @param {AnchorPosition} anchorPosition
+   * @return {WorkingBucket}
+   */
+  function _newBucket(anchorPosition) {
```

Comment:
Can you remove the underscore prefix here. Perhaps it was needed in an earlier iteration of the code but it isn't needed any longer

This helper could be moved to the top level of the module as it doesn't capture any variables - I leave that up to you.

Reply:
I'll kill the underscore, no prob. I'd like to keep this function here as it's only relevant to the iteration beneath it and I don't want a future dev to have to do too much searching through the module to find it.